ddtrace/internal/rate_limiter.py

```diff
@@ -0,0 +1,79 @@
+import time
+import threading
+
+
+class RateLimiter(object):
+    """
+    A token bucket rate limiter implementation
+    """
+    __slots__ = ('rate_limit', 'tokens', 'max_tokens', 'last_update', '_lock')
+
+    def __init__(self, rate_limit):
+        """
+        Constructor for RateLimiter
+
+        :param rate_limit: The rate limit to apply for number of transactions per second.
+            rate limit > 0 max number of transactions to allow per second,
+            rate limit == 0 to disallow any transactions,
+            rate limit < 0 to disable any rate limiting
+        :type rate_limit: :obj:`int`
+        """
+        self.rate_limit = rate_limit
+        self.tokens = rate_limit
+        self.max_tokens = rate_limit
+
+        self.last_update = time.time()
+        self._lock = threading.Lock()
+
+    def is_allowed(self):
+        """
+        Check whether the current transaction is allowed or not
+
+        This method will also reduce the number of available transactions by 1
+
+        :returns: Whether the current transaction is allowed or not
+        :rtype: :obj:`bool`
+        """
+        # Lock, we need this to be thread safe, it should be shared by all threads
+        with self._lock:
```

Comment:
IIUC the only thing that might change and needs locking is code around `self.tokens`. It looks to me this lock is too wide. For example, if `rate_limit == 0` there's no need to lock the lock at all.